src/binarylane/config/sources.py

```diff
@@ -0,0 +1,130 @@
+from __future__ import annotations
+
+import argparse
+import configparser
+import os
+import sys
+from abc import ABC
+from pathlib import Path
+from typing import ClassVar, Dict, Optional
+
+
+class _SourceBase(ABC):
+    _config: Dict[str, str]
+
+    def get(self, name: str) -> Optional[str]:
+        return self._config.get(name, None)
+
+
+class CommandlineSource(_SourceBase):
+    def __init__(self, config: argparse.Namespace) -> None:
+        self._config = {key.replace("_", "-"): str(value) for key, value in vars(config).items() if value is not None}
+
+
+class DefaultSource(_SourceBase):
+    def __init__(self) -> None:
+        self._config = {
+            "api-url": "https://api.binarylane.com.au",
+            "context": "bl",
+        }
+
+
+class EnvironmentSource(_SourceBase):
+    prefix: ClassVar[str] = "BL_"
+
+    def __init__(self) -> None:
+        self._config = {self._get_name(key): value for key, value in os.environ.items() if key.startswith(self.prefix)}
+
+    def _get_name(self, key: str) -> str:
+        return key[len(self.prefix) :].lower().replace("_", "-")
+
+
+class FileSource:
+    DEFAULT_PATH = Path("__DEFAULT__")
+    _DIRNAME = "binarylane"
+    _FILENAME = "config.ini"
+    _path: Optional[Path]
+    section_name: str
+
+    _parser: configparser.ConfigParser
+
+    def __init__(self, config_file: Optional[Path] = None) -> None:
+        if config_file is self.DEFAULT_PATH:
+            config_file = self._get_config_dir() / self._FILENAME
+        self._path = config_file
+
+        self._parser = configparser.ConfigParser()
```

Comment:
By default config parser supports basic interpolation, it would be good for the first version to opt out of interpolation with
`interpolation=None` so that a future rewrite in a different language only needs to support the basic INI format.

Reply:
Agreed, I will review the other args the constructor accepts in case there's any other features we may not want currently.